rust/rebase/src/flow/poap_ownership.rs

```diff
@@ -0,0 +1,193 @@
+            ));
+        }
+
+        let now = Utc::now();
+        let then = DateTime::parse_from_rfc3339(timestamp)
+            .map_err(|e| FlowError::Validation(e.to_string()))?;
+
+        if then > now {
+            return Err(FlowError::Validation(format!(
+                "Timestamp provided comes from the future"
+            )));
+        }
+
+        if now - Duration::minutes(self.max_elapsed_minutes) > then {
+            return Err(FlowError::Validation(format!(
+                "Validation window has expired"
+            )));
+        };
+        Ok(())
+    }
+}
+
+#[derive(Deserialize, Serialize)]
+struct PoapResEntry {
+    event: PoapEventEntry,
+    #[serde(rename = "tokenId")]
+    token_id: String,
+    owner: String,
+    chain: String,
+    // NOTE: This date is in the format "YYYY-MM-DD HH-mm-ss"
+    created: String,
+}
+
+#[derive(Deserialize, Serialize)]
+struct PoapEventEntry {
+    id: u64,
+    fancy_id: String,
+    name: String,
+    event_url: String,
+    image_url: String,
+    country: String,
+    city: String,
+    description: String,
+    year: u64,
+    // NOTE: These dates are in the format "DD-MonthShortName-YYYY"
+    start_date: String,
+    end_date: String,
+    // TODO: Test for this?
+    expiry_date: String,
+    supply: u64,
+}
+
+#[async_trait(?Send)]
+impl Flow<Ctnt, Stmt, Prf> for PoapOwnership {
+    fn instructions(&self) -> Result<Instructions, FlowError> {
+        Ok(Instructions {
+            statement: "Enter the event id of the POAP you want to verify ownership of."
+                .to_string(),
+            signature: "Sign a statement attesting to ownership of the POAP.".to_string(),
+            witness: "Send the attestation and the signature to the witness and issue a credential"
+                .to_string(),
+            statement_schema: schema_for!(Stmt),
+            witness_schema: schema_for!(Prf),
+        })
+    }
+
+    async fn statement<I: Issuer>(
+        &self,
+        stmt: &Stmt,
+        _issuer: &I,
+    ) -> Result<FlowResponse, FlowError> {
+        self.sanity_check(&stmt.issued_at)?;
+
+        // TODO: Investigate!
+        // Can POAPs be attached to non EIP155 DIDs?
+        match stmt.subject {
+            Subjects::Pkh(Pkh::Eip155(_)) => {}
+            _ => {
+                return Err(FlowError::Validation(
+                    "Currently only supports Ethereum Addresses for POAP Ownership flow"
+                        .to_string(),
+                ))
+            }
+        }
+
+        Ok(FlowResponse {
+            statement: stmt.generate_statement()?,
+            delimitor: None,
+        })
+    }
+
+    async fn validate_proof<I: Issuer>(&self, proof: &Prf, _issuer: &I) -> Result<Ctnt, FlowError> {
+        self.sanity_check(&proof.statement.issued_at)?;
+
+        let u = Url::parse(&format!(
+            "https://api.poap.tech/actions/scan/{}",
```

Comment:
Is there just one poap api endpoint? if not, should this be an env variable/`self.api`? also probably a concern for later

Reply:
For the purposes of the demo app, there's only one API endpoint, and there's only one official POAP API , so probably?

That's definitely something I plan to look into when I circle around post-demo -- if it's one use a const, if it's many use a `self.api`, like you suggest.